### PR TITLE
Refactor ESLint rules & fix transpile errors when optional dependencies are not installed

### PR DESCRIPTION
**Pull request by Nuckyz**

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -9,6 +347,8 @@
 				"SwitchCase": 1
 			}
 		],
-		"jsdoc/newline-after-description": ["error", "never"]
+		"jsdoc/newline-after-description": ["error", "never"],
+		"@typescript-eslint/ban-ts-comment": "off",
+		"@typescript-eslint/consistent-type-imports": "off"
```
Comment (DonovanDMC) [inline .eslintrc.json]:
```suggestion
		"@typescript-eslint/consistent-type-imports": [
			"error",
			{
				"prefer": "type-imports",
				"disallowTypeAnnotations": true
			}
		]
```

I don't want either of these off

File: lib/Client.ts [hunk 2]
```diff
@@ -21,6 +21,7 @@ import type StageChannel from "./structures/StageChannel";
 import Util from "./util/Util";
 
 /* eslint-disable */
+/* @ts-ignore */
```
Comment (DonovanDMC) [inline lib/Client.ts]:
```suggestion
// @ts-ignore
```

File: lib/gateway/Shard.ts [hunk 3]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
 import type { Data } from "ws";
 import { WebSocket } from "ws";
 import type Pako from "pako";
+/* @ts-ignore */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```

File: lib/gateway/Shard.ts [hunk 4]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
 import type { Inflate } from "zlib-sync";
 import { randomBytes } from "crypto";
 import { inspect } from "util";
 import assert from "assert";
 
-/* eslint-disable */
+/* @ts-ignore */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```

File: lib/gateway/Shard.ts [hunk 5]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
 let Erlpack: typeof import("erlpack") | undefined;
 try {
+    /* eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
```

File: lib/gateway/Shard.ts [hunk 6]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
     Erlpack = require("erlpack");
 } catch { }
+/* @ts-ignore */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```

File: lib/gateway/Shard.ts [hunk 7]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
 let ZlibSync: typeof import("pako") | typeof import("zlib-sync") | undefined, zlibConstants: typeof import("pako").constants | typeof import("zlib-sync") | undefined;
 try {
-    ZlibSync = require("zlib-sync");
-    zlibConstants = require("zlib-sync");
+    ZlibSync = require("zlib-sync"); /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
    ZlibSync = require("zlib-sync");
```

File: lib/gateway/Shard.ts [hunk 8]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
+    zlibConstants = require("zlib-sync"); /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
    zlibConstants = require("zlib-sync");
```

File: lib/gateway/Shard.ts [hunk 9]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
 } catch {
     try {
-        ZlibSync = require("pako");
-    zlibConstants = require("pako").constants;
+        ZlibSync = require("pako"); /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
        ZlibSync = require("pako");
```

File: lib/gateway/Shard.ts [hunk 10]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
+        zlibConstants = require("pako").constants; /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
        zlibConstants = require("pako").constants;
```

File: lib/gateway/Shard.ts [hunk 11]
```diff
@@ -188,10 +190,10 @@ export default class Shard extends TypedEmitter<ShardEvents> {
             this.ws = new WebSocket(this.client.gatewayURL, this.client.shards.options.ws);
         }
 
-        this.ws.on("close", this.onWSClose);
-        this.ws.on("error", this.onWSError);
-        this.ws.on("message", this.onWSMessage);
-        this.ws.on("open", this.onWSOpen);
+        this.ws.on("close", this.onWSClose); /* eslint-disable-line @typescript-eslint/unbound-method */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        /* eslint-disable @typescript-eslint/unbound-method */
        this.ws.on("close", this.onWSClose);
```

File: lib/gateway/Shard.ts [hunk 12]
```diff
@@ -188,10 +190,10 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
+        this.ws.on("error", this.onWSError); /* eslint-disable-line @typescript-eslint/unbound-method */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        this.ws.on("error", this.onWSError);
```

File: lib/gateway/Shard.ts [hunk 13]
```diff
@@ -188,10 +190,10 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
+        this.ws.on("message", this.onWSMessage); /* eslint-disable-line @typescript-eslint/unbound-method */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        this.ws.on("message", this.onWSMessage);
```

File: lib/gateway/Shard.ts [hunk 14]
```diff
@@ -188,10 +190,10 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
+        this.ws.on("open", this.onWSOpen); /* eslint-disable-line @typescript-eslint/unbound-method */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
        this.ws.on("open", this.onWSOpen);
        /* eslint-enable @typescript-eslint/unbound-method */
```

File: lib/gateway/Shard.ts [hunk 15]
```diff
@@ -1064,7 +1066,7 @@ export default class Shard extends TypedEmitter<ShardEvents> {
                 break;
             }
 
-            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+            /* eslint-disable-next-line @typescript-eslint/restrict-template-expressions */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
```

File: lib/gateway/Shard.ts [hunk 16]
```diff
@@ -56,28 +56,29 @@ import { is } from "../util/Util";
...
-        ZlibSync = require("pako");
-    zlibConstants = require("pako").constants;
+        ZlibSync = require("pako"); /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment */
+        zlibConstants = require("pako").constants; /* eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
     } catch {}
 }
-/* eslint-enable */
-
 
 /* eslint-disable @typescript-eslint/unbound-method */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
```

File: lib/gateway/Shard.ts [hunk 17]
```diff
@@ -175,6 +176,7 @@ export default class Shard extends TypedEmitter<ShardEvents> {
         if (this.client.shards.options.compress) {
             if (!ZlibSync) throw new Error("Cannot use compression without pako or zlib-sync.");
             this.client.emit("debug", "Initializing zlib-sync-based compression");
+            /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
```

File: lib/gateway/Shard.ts [hunk 18]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
             assert(is<Buffer>(data));
             if (this.client.shards.options.compress) {
                 if (data.length >= 4 && data.readUInt32BE(data.length - 4) === 0xFFFF) {
+                    /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```

File: lib/gateway/Shard.ts [hunk 19]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                     this.#sharedZLib.push(data, zlibConstants!.Z_SYNC_FLUSH);
+                    /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```

File: lib/gateway/Shard.ts [hunk 20]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                     if (this.#sharedZLib.err) {
+                        /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/restrict-template-expressions */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/restrict-template-expressions
```

File: lib/gateway/Shard.ts [hunk 21]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                         this.client.emit("error", new Error(`zlib error ${this.#sharedZLib.err}: ${this.#sharedZLib.msg || ""}`));
                         return;
                     }
 
+                    /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-argument */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-argument
```

File: lib/gateway/Shard.ts [hunk 22]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                     data = Buffer.from(this.#sharedZLib.result || "");
                     if (Erlpack) {
+                        /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```

File: lib/gateway/Shard.ts [hunk 23]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                         return this.onPacket(Erlpack.unpack(data as Buffer) as AnyReceivePacket);
                     } else {
                         return this.onPacket(JSON.parse(data.toString()) as AnyReceivePacket);
                     }
                 } else {
+                    /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```

File: lib/gateway/Shard.ts [hunk 24]
```diff
@@ -1186,22 +1188,29 @@ export default class Shard extends TypedEmitter<ShardEvents> {
...
                     this.#sharedZLib.push(data, false);
                 }
             } else if (Erlpack) {
+                /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```

File: lib/gateway/Shard.ts [hunk 25]
```diff
@@ -1443,6 +1452,7 @@ export default class Shard extends TypedEmitter<ShardEvents> {
             let i = 0, waitFor = 1;
             const func = () => {
                 if (++i >= waitFor && this.ws && this.ws.readyState === WebSocket.OPEN) {
+                    /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */
```
Comment (DonovanDMC) [inline lib/gateway/Shard.ts]:
```suggestion
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
```

File: .eslintrc.json [hunk 26]
```diff
@@ -1,6 +1,344 @@
...
+						"static-method",
+						"private-method",
+						"protected-method",
+						"private-get",
+						"protected-get",
+						"get",
+						"private-set",
+						"protected-set",
+						"set",
+						"method"
```
Comment (DonovanDMC) [inline .eslintrc.json]:
```suggestion
						"signature",
						"private-static-field",
						"protected-static-field",
						"static-field",
						"private-field",
						"protected-field",
						"field",
						"abstract-field",
						"private-abstract-method",
						"protected-abstract-method",
						"abstract-method",
						"private-constructor",
						"protected-constructor",
						"constructor",
						"private-static-method",
						"protected-static-method",
						"static-method",
						"private-method",
						"protected-method",
						"private-get",
						"protected-get",
						"get",
						"private-set",
						"protected-set",
						"set",
						"method"
```

File: .eslintrc.json [hunk 27]
```diff
@@ -1,6 +1,344 @@
...
+					],
+					"order": "alphabetically"
+				}
+			}
+		],
 		"@typescript-eslint/no-unnecessary-type-assertion": "off",
```
Comment (DonovanDMC) [inline .eslintrc.json:342]:
```suggestion
		"@typescript-eslint/no-unnecessary-type-assertion": "error",
```

**Review by DonovanDMC — CHANGES_REQUESTED** [review]

File: lib/Client.ts [hunk 28]
```diff
@@ -21,6 +21,7 @@ import type StageChannel from "./structures/StageChannel";
 import Util from "./util/Util";
 
 /* eslint-disable */
```
Comment (DonovanDMC) [inline lib/Client.ts]: replace this with a set of disables, like you did in shard

Comment (Nuckyz) [reply lib/Client.ts]: all done

**Review by DonovanDMC — APPROVED** [review]

**Pull request merged**